### DISCUSSION TITLE
Apply EnvironmentDict to another and Rex confusion.

### DISCUSSION THREAD

**Discussion by herronelou** [AcademySoftwareFoundation/rez#1355] [Q&A] [answered]
Hello.

I'm digging a little bit into rez.rex, and I'm having a hard time wrapping my head around it. Maybe I'm not even looking in the right place so I'll explain my main problem, then what I was currently thinking.

**Main issue:**
We are looking for a way to apply our own environment variables on top of the rez environment variables from our launcher.
Our launcher is using a ResolvedContext().execute_shell() with a post_actions_callback. 
To this post action callback, I can pass a dict as argument and have the callback update the env based on that dict.
The problem is that now we have no control whether the variable should be set or appended, like we do when using the env objects in rez.

**My thinking:**
I could use rez native EnvironmentDict object (or RexExecutor) in my config files to set all the envs, and pass this object to my callback.
I noticed the RexExecutor stores all the actions that were done, and that the callback does give me access to the executor. I thought maybe I can create an executor to which we append/set envs, which would get stored as actions, and somehow I would later be able to re-apply these actions to the main executor, however I can't find a way to do anything with the stored actions.
I'm having a hard time expressing what I'm trying to do with words, so here is some pseudo-code of what I was thinking:

```
my_config = RexExecutor()
my_config.env.SOME_PATH.append('foo')
my_config.env.BAR = 'bar'

def my_callback(main_executor, second_executor):
    for action in second_executor.actions:
        main_executor.do_action(action)  # This is what I can't seem to find how to do, is it even possible?

rc = ResolvedContext(my_request)
rc.execute_shell(command=baz, post_actions_callback=partial(my_callback, second_executor=my_config)
```

This way, whether an env was appended/prepended/set/unset would be respected.

Am I completely looking in the wrong place?

**Comment by nerdvegas** [comment, marked as answer]
It might be possible to take that kind of approach, but honestly I think you're overthinking it! How about -

```
env_updates = {
    "SOME_PATH": ("append", "foo"),
    "BAR": ("set", "bar")
}

def my_callback(ex):
    for varname, (op, val) in env_update.items():
        if op == "append":
            ex.env[varname].append(val)
        else if ...:
            ...  # and so on
```

I'm sure we could improve the rex API to support things like playing the actions of one Executor onto another, but tbh this kind of thing is not the primary goal of rez and I don't think we'd benefit from spending resources on that. As its own project though - ie a python-based, shell-agnostic language that can transpile(?) to any number of target shell languages - it would absolutely make sense do to this kind of work. Rex itself could definitely be its own project, but that's not where we are presently.

**Reply by herronelou** [reply]
You are right, I'm most likely overthinking this.
This had a certain "elegance" to being able to write our extra env configs the same way we write rez packages, but if it's not meant for that I can use something else.
I'll probably end up doing something similar to what you propose, but wrapped into a class so that I can make it look like the same syntax used by rez.
